homeassistant/components/climate/modbus.py

```diff
@@ -0,0 +1,182 @@
+    data_type = config.get(CONF_DATA_TYPE)
+    count = config.get(CONF_COUNT)
+    precision = config.get(CONF_PRECISION)
+
+    add_devices([ModbusThermostat(hass, name, modbus_slave,
+                                  target_temp_register, current_temp_register,
+                                  data_type, count, precision)], True)
+
+
+class ModbusThermostat(ClimateDevice):
+    """Representation of a Modbus Thermostat."""
+
+    def __init__(self, hass, name, modbus_slave, target_temp_register,
+                 current_temp_register, data_type, count, precision):
+        """Initialize the unit."""
+        self.hass = hass
+        self._name = name
+        self._slave = modbus_slave
+        self._target_temperature_register = int(target_temp_register)
+        self._current_temperature_register = int(current_temp_register)
+        self._target_temperature = None
+        self._current_temperature = None
+        self._data_type = data_type
+        self._count = int(count)
+        self._precision = precision
+        self._structure = '>f'
+        self._error = False
+
+        data_types = {DATA_TYPE_INT: {1: 'h', 2: 'i', 4: 'q'}}
+        data_types[DATA_TYPE_UINT] = {1: 'H', 2: 'I', 4: 'Q'}
+        data_types[DATA_TYPE_FLOAT] = {1: 'e', 2: 'f', 4: 'd'}
+
+        self._structure = '>{}'.format(data_types[self._data_type]
+                                       [self._count])
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        return SUPPORT_FLAGS
+
+    def update(self):
+        """Update Target & Current Temperature."""
+        try:
+            result = modbus.HUB.read_holding_registers(
+                self._slave, self._target_temperature_register, self._count)
+        except AttributeError as ex:
+            self._error = True
+            _LOGGER.error(ex)
+
+        byte_string = b''.join(
+            [x.to_bytes(2, byteorder='big') for x in result.registers])
+        val = struct.unpack(self._structure, byte_string)[0]
+        formatted_string = format(val, '.{}f'.format(self._precision))
+        self._target_temperature = float(formatted_string)
+
+        try:
+            result = modbus.HUB.read_holding_registers(
+                self._slave, self._current_temperature_register, self._count)
+        except AttributeError as ex:
+            self._error = True
+            _LOGGER.error(ex)
+
+        byte_string = b''.join(
+            [x.to_bytes(2, byteorder='big') for x in result.registers])
+        val = struct.unpack(self._structure, byte_string)[0]
+        formatted_string = format(val, '.{}f'.format(self._precision))
+        self._current_temperature = float(formatted_string)
+
+    @property
+    def name(self):
+        """Return the name of the climate device."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the current state - implied."""
+        if self._error:
```

Comment:
is the logic right ?

Reply:
Good catch - thanks!

Reply:
Improving of this property:
Can't it be one of the "climate" states ?

(homeassistant/components/climate/__init__.py)

Reply:
The only problem with that is I hoped to keep the functionality simple, since Modbus is just a communications protocol. The controllers which implement this protocol vary wildly. Depending on the manufacturer of the controller, it may or may not support a register that has an output state (heating, cooling, off). If there's requests for more features from users of this module I would be happy to submit them. It's just not required for the controller to function. I was asked about this in an older PR I closed.